src/main/scala/chisel3/aop/Select.scala

```diff
@@ -0,0 +1,423 @@
+    * @param module Module to collect components, as well as all children module it directly and indirectly instantiates
+    * @param collector Collector partial function to pick, given a module, which components to collect
+    * @param tag Required for generics to work, should ignore this
+    * @tparam T Type of the component that will be collected
+    * @return
+    */
+  def collectDeep[T](module: BaseModule)(collector: PartialFunction[BaseModule, T])(implicit tag: TypeTag[T]): Iterable[T] = {
+    check(module)
+    val myItems = collector.lift(module)
+    val deepChildrenItems = instances(module).flatMap {
+      i => collectDeep(i)(collector)
+    }
+    myItems ++ deepChildrenItems
+  }
+
+  /** Selects all instances directly instantiated within given module
+    * @param module
+    * @return
+    */
+  def instances(module: BaseModule): Seq[BaseModule] = {
+    check(module)
+    module._component.get.asInstanceOf[DefModule].commands.collect {
+      case i: DefInstance => i.id
+    }
+  }
+
+  /** Selects all registers directly instantiated within given module
+    * @param module
+    * @return
+    */
+  def registers(module: BaseModule): Seq[Data] = {
+    check(module)
+    module._component.get.asInstanceOf[DefModule].commands.collect {
+      case r: DefReg => r.id
+      case r: DefRegInit => r.id
+    }
+  }
+
+  /** Selects all ios directly contained within given module
+    * @param module
+    * @return
+    */
+  def ios(module: BaseModule): Seq[Data] = {
+    check(module)
+    module._component.get.asInstanceOf[DefModule].ports.map(_.id)
+  }
+
+  /** Selects all SyncReadMems directly contained within given module
+    * @param module
+    * @return
+    */
+  def syncReadMems(module: BaseModule): Seq[SyncReadMem[_]] = {
+    check(module)
+    module._component.get.asInstanceOf[DefModule].commands.collect {
+      case r: DefSeqMemory => r.id.asInstanceOf[SyncReadMem[_]]
+    }
+  }
+
+  /** Selects all Mems directly contained within given module
+    * @param module
+    * @return
+    */
+  def mems(module: BaseModule): Seq[Mem[_]] = {
+    check(module)
+    module._component.get.asInstanceOf[DefModule].commands.collect {
+      case r: DefMemory => r.id.asInstanceOf[Mem[_]]
+    }
+  }
+
+  /** Selects all arithmetic or logical operators directly instantiated within given module
+    * @param module
+    * @return
+    */
+  def ops(module: BaseModule): Seq[(String, Data)] = {
+    check(module)
+    module._component.get.asInstanceOf[DefModule].commands.collect {
+      case d: DefPrim[_] => (d.op.name, d.id)
+    }
+  }
+
+  /** Selects a kind of arithmetic or logical operator directly instantiated within given module
+    * The kind of operators are contained in [[chisel3.internal.firrtl.PrimOp]]
+    * @param opKind the kind of operator, e.g. "mux", "add", or "bits"
+    * @param module
+    * @return
+    */
+  def ops(opKind: String)(module: BaseModule): Seq[Data] = {
```

Comment:
Should this use a `String` parameter or `chisel3.internal.firrtl.PrimOp`? I expect your motivation here is to not rely on `chisel3.internal.firrtl`. I think that's fine, just wanted to clarify.

Reply:
Yeah, I want to keep the `internal.firrtl` stuff as hidden away as possible.